explorer/src/db/mod.rs

```diff
@@ -0,0 +1,848 @@
+            if let Some(b) = state_ref.state().blocks.lookup(block_id) {
+                block = block.or_else(|| Some(Arc::clone(b)));
+                tips.push((hash, state_ref));
+            }
+        }
+
+        block.map(|b| (b, tips))
+    }
+
+    pub async fn get_epoch(&self, epoch: Epoch) -> Option<EpochData> {
+        let tips = self.multiverse.tips().await;
+        let (_, state_ref) = &tips[0];
+
+        state_ref
+            .state()
+            .epochs
+            .lookup(&epoch)
+            .map(|e| e.as_ref().clone())
+    }
+
+    pub async fn is_block_confirmed(&self, block_id: &HeaderHash) -> bool {
+        let current_branch = self
+            .multiverse
+            .get_ref(&self.longest_chain_tip.get_block_id().await)
+            .await
+            .unwrap();
+
+        if let Some(block) = current_branch.state().blocks.lookup(block_id) {
+            let confirmed_block_chain_length: ChainLength = self
+                .stable_store
+                .confirmed_block_chain_length
+                .load(Ordering::Acquire)
+                .into();
+            block.chain_length <= confirmed_block_chain_length
+        } else {
+            false
+        }
+    }
+
+    pub async fn find_blocks_by_chain_length(&self, chain_length: ChainLength) -> Vec<HeaderHash> {
+        let mut hashes = Vec::new();
+
+        for (_hash, state_ref) in self.multiverse.tips().await.iter() {
+            if let Some(hash) = state_ref.state().chain_lengths.lookup(&chain_length) {
+                hashes.push(**hash);
+            }
+        }
+
+        hashes.sort_unstable();
+        hashes.dedup();
+
+        hashes
+    }
+
+    pub async fn find_blocks_by_transaction(&self, transaction_id: &FragmentId) -> Vec<HeaderHash> {
+        let mut txs: Vec<_> = self
+            .multiverse
+            .tips()
+            .await
+            .iter()
+            .filter_map(|(_tip_hash, state_ref)| {
+                state_ref
+                    .state()
+                    .transactions
+                    .lookup(transaction_id)
+                    .map(|arc| *arc.clone())
+            })
+            .collect();
+
+        txs.sort_unstable();
+        txs.dedup();
+
+        txs
+    }
+
+    pub async fn get_stake_pool_blocks(
+        &self,
+        pool: &PoolId,
+    ) -> Option<Arc<PersistentSequence<HeaderHash>>> {
+        let pool = pool.clone();
+
+        // this is a tricky query, one option would be to take a hash and return
+        // only the blocks from a particular branch, but it's not like a stake
+        // pool would produce inconsistent branches itself, although there may
+        // be a need to know the blocks that a stake pool got in the main branch
+        // too.
+        // for the time being, this query uses the maximum, because the branch
+        // that has more blocks from this particular stake pool has all the
+        // blocks produced by it
```

Comment:
> and the main chain will likely hold the most blocks produced by any stake pool

I would expect the losing block to be mined on top of the main chain then, though (minus the winner block). That said, this query probably needs some tweaking.

Reply:
> I would expect the losing block to be mined on top of the main chain then

I don't get this, it will be a different block with likely different transactions and in a different slot (which means the stake pool would have to win another lottery to produce it, which is not guaranteed)

Reply:
I mean, the actual block doesn't matter, the parent is. 

But I think you are right. Maybe picking the branch with the greatest `ChainLength` produced by that stake pool would have been more accurate. But maybe this query should just ask for a hash of the state that you want to use for the query (like the current tip).

I think it may be better to extract this to an issue instead of changing it here, though. I did change some things like the treasury because I had to, but I was trying to avoid changing much behavior in this PR, as it becomes harder to see.